File: snippets/cpp/graph/hierholzer.cpp
vi eulerian_path(vector<vpii> &g, int s = 0) {
    int n = sza(g); vector<vi> used(n);
    for (int u = 0; u < n; u++)
        used[u].resize(sza(g[u]));
    vi path; stack<int> st; st.push(s);
    while (!st.empty()) {
        int u = st.top();
        bool found = false;
        for (int i = 0; i < sza(g[u]); i++) {
            if (!used[u][i]) {
                used[u][i] = 1;
                st.push(g[u][i].f);
                found = true;
                break;
            }
        }
        if (!found) { path.pb(u); st.pop(); }
    }
    reverse(all(path));
    return path;
}

File: snippets/rust/graph/hierholzer.rs
// Eulerian path/circuit via Hierholzer's algorithm.
// `g[u]` is a list of (v, edge_id) outgoing edges; for undirected graphs add the edge twice
// with the same `edge_id` so it can be marked used from either side.
fn eulerian_path(g: &Vec<Vec<(usize, usize)>>, s: usize, num_edges: usize) -> Vec<usize> {
    let n = g.len();
    let mut idx = vec![0usize; n];
    let mut used = vec![false; num_edges];
    let mut path = Vec::new();
    let mut st = vec![s];
    while let Some(&u) = st.last() {
        while idx[u] < g[u].len() && used[g[u][idx[u]].1] { idx[u] += 1; }
        if idx[u] == g[u].len() {
            path.push(u);
            st.pop();
        } else {
            let (v, eid) = g[u][idx[u]];
            used[eid] = true;
            idx[u] += 1;
            st.push(v);
        }
    }
    path.reverse();
    path
}
